src/menhir.ml

```diff
@@ -102,31 +145,114 @@ module Run (P : PARAMS) = struct
     | Some _ ->
         [ stanza ]
 
-  (* [process stanza] converts a Menhir stanza into a rule, which it installs. *)
+  (* ------------------------------------------------------------------------ *)
 
-  (* Reminder (from arg_spec.mli):
+  (* The [--infer-*] commands should not be passed by the user; we take care
+     of using these commands appropriately. Fail if they are present. *)
 
-     [Deps]           is for command line arguments that are dependencies.
-     [As]             is for command line arguments
-                      that are neither dependencies nor targets.
-     [Hidden_targets] is for targets that are *not* command line arguments.
+  let () =
+    List.iter stanzas ~f:(fun (stanza : stanza) ->
+      List.iter stanza.flags ~f:(fun flag ->
+        match flag with
+        | "--depend"
+        | "--raw-depend"
+        | "--infer"
+        | "--infer-write-query"
+        | "--infer-read-reply" ->
+            Loc.fail stanza.loc
```

Comment:
Slightly nicer would be to use a more precise location for the flags. For example, here's what you can do attach a loc to every flag:

```
diff --git a/src/jbuild.ml b/src/jbuild.ml
index c1717d63..dd9e3a3d 100644
--- a/src/jbuild.ml
+++ b/src/jbuild.ml
@@ -1077,7 +1077,7 @@ end
 module Menhir = struct
   type t =
     { merge_into : string option
-    ; flags      : string list
+    ; flags      : (Loc.t * string) list
     ; modules    : string list
     ; mode       : Rule.Mode.t
     ; loc        :  Loc.t
@@ -1086,7 +1086,7 @@ module Menhir = struct
   let v1 =
     record
       (field_o "merge_into" string >>= fun merge_into ->
-       field "flags" (list string) ~default:[] >>= fun flags ->
+       field "flags" (list (located string)) ~default:[] >>= fun flags ->
        field "modules" (list string) >>= fun modules ->
        Rule.Mode.field >>= fun mode ->
        return
diff --git a/src/jbuild.mli b/src/jbuild.mli
index 609ccece..a0a60e98 100644
--- a/src/jbuild.mli
+++ b/src/jbuild.mli
@@ -298,7 +298,7 @@ end
 module Menhir : sig
   type t =
     { merge_into : string option
-    ; flags      : string list
+    ; flags      : (Loc.t * string) list
     ; modules    : string list
     ; mode       : Rule.Mode.t
     ; loc        : Loc.t
```

And if that is a bit fine grained, we can consider `Loc.t * (string list)`. But it's really nice to have precise error messages for your editor to jump to.

Reply:
Regarding `make test`: this command fails with several errors in `upstream/master`. Am I supposed to verify that my PR does not add *new* failures?